core/Execute.cpp

```diff
@@ -1,163 +1,69 @@
-
-        ex_inst->setStatus(Inst::Status::COMPLETED);
-        // We're not busy anymore
-        unit_busy_ = false;
-
-        // Count the instruction as completely executed
-        ++total_insts_executed_;
-
-        // Schedule issue if we have instructions to issue
-        if (ready_queue_.size() > 0) {
-            issue_inst_.schedule(sparta::Clock::Cycle(0));
-        }
-    }
-
-    void Execute::flushInst_(const FlushManager::FlushingCriteria & criteria)
+    void ExecuteFactory::bindLate (sparta::TreeNode *node)
     {
-        if(SPARTA_EXPECT_FALSE(info_logger_)) {
-            info_logger_ << "Got flush for criteria: " << criteria;
-        }
-
-        // Flush instructions in the ready queue
-        ReadyQueue::iterator it = ready_queue_.begin();
-        uint32_t credits_to_send = 0;
-        while(it != ready_queue_.end()) {
-            if((*it)->getUniqueID() >= uint64_t(criteria)) {
-                ready_queue_.erase(it++);
-                ++credits_to_send;
-            }
-            else {
-                ++it;
-            }
-        }
-        if(credits_to_send) {
-            out_scheduler_credits_.send(credits_to_send, 0);
-        }
-
-        // Cancel outstanding instructions awaiting completion and
-        // instructions on their way to issue
-        auto cancel_critera = [criteria](const InstPtr & inst) -> bool {
-            if(inst->getUniqueID() >= uint64_t(criteria)) {
-                return true;
-            }
-            return false;
-        };
-        complete_inst_.cancelIf(cancel_critera);
-        issue_inst_.cancel();
-
-        if(complete_inst_.getNumOutstandingEvents() == 0) {
-            unit_busy_ = false;
-            collected_inst_.closeRecord();
+        auto bind_ports =
+            [node] (const std::string & left, const std::string & right) {
+                sparta::bind(node->getParent()->getChildAs<sparta::Port>(left),
+                             node->getParent()->getChildAs<sparta::Port>(right));
+            };
+
+        // Bind all of the pipes created to Dispatch and Flushmanager
+        for(auto & exe_pipe_rtn : exe_pipe_tns_)
+        {
+            const std::string exe_credits_out = "execute." + exe_pipe_rtn->getName() +
+                ".ports.out_scheduler_credits";
+            const std::string disp_credits_in = "dispatch.ports.in_" + exe_pipe_rtn->getName() + "_credits";
+            bind_ports(exe_credits_out, disp_credits_in);
+
+            const std::string exe_inst_in   = "execute." + exe_pipe_rtn->getName() + ".ports.in_execute_write";
+            const std::string disp_inst_out = "dispatch.ports.out_" + exe_pipe_rtn->getName() + "_write";
+            bind_ports(exe_inst_in, disp_inst_out);
+
+            const std::string exe_flush_in = "execute." +  exe_pipe_rtn->getName() + ".ports.in_reorder_flush";;
+            const std::string flush_manager = "flushmanager.ports.out_retire_flush";
+            bind_ports(exe_flush_in, flush_manager);
```

Comment:
Hm, this is something I'm always wondering about -- bind these units here or at topology level? One concern I come up with is -- that would make the execution unit test have to create dispatch instance.

Reply:
This is true -- I do "hard-code" dispatch and the flush manager in this factory.  I don't like that either.  Perhaps moving this to the topology would make sense...

Reply:
Done -- moved to CPUTopology.